pallets/protos/src/lib.rs

```diff
@@ -531,8 +529,69 @@ pub mod pallet {
 	}
 
 	impl<T: Config> Pallet<T> {
+		/// This function is deprecated and will never be used
 		pub fn get_by_tag(tags: Tags) -> Option<Vec<Hash256>> {
-			<ProtosByTag<T>>::get(&tags)
+			Some(<ProtosByTag<T>>::iter_key_prefix(tags).collect::<Vec<Hash256>>())
+		}
+
+		fn is_proto_having_any_tags(proto_hash: &Hash256, tags: &Vec<Tags>) -> bool {
+			if let Some(struct_proto) = <Protos<T>>::get(proto_hash) {
+				tags.into_iter().any(|tag| struct_proto.tags.contains(&tag))
+			} else {
+				false
+			}
+		}
+
+		pub fn get_by_tags(tags: Vec<Tags>, owner: Option<T::AccountId>, limit: u32) -> Vec<Hash256> {
+
+
+			// log::info!("inside get_by_tags");
+			// log::info!("tags: {:?}, owner: {:?}, limit: {}", tags, owner, limit);
+
+			match owner {
+				Some(owner) => {
+
+					let iter_protos = <ProtosByOwner<T>>::iter_key_prefix(ProtoOwner::User(owner));
+
+					let iter_protos_filtered = iter_protos.filter(|proto| Self::is_proto_having_any_tags(proto, &tags));
+					let iter_protos_limited = iter_protos_filtered.take(limit as usize);
+					iter_protos_limited.collect::<Vec<Hash256>>()
+
+				},
+				None => {
+					let mut remaining = limit;
+
+					let iter_protos = tags.into_iter().map(|tag| {
+						let vector_protos = <ProtosByTag<T>>::iter_key_prefix(tag).take(remaining as usize).collect::<Vec<Hash256>>();
+						remaining -= vector_protos.len() as u32;
+						vector_protos
+					}).flatten().take(limit as usize);
```

Comment:
we apply the limit too late, if there are a million of protos we would get the full million here no?

Reply:
We also apply the limit inside the `map` function (`<ProtosByTag<T>>::iter_key_prefix(tag).take(remaining as usize).collect::<Vec<Hash256>>();`). 

So I don't think so? Please let me know if I'm wrong